hail/python/hail/expr/aggregators/aggregators.py

```diff
@@ -187,6 +187,43 @@ def _check_agg_bindings(expr, bindings):
         raise ExpressionException("dynamic variables created by 'hl.bind' or lambda methods like 'hl.map' may not be aggregated")
 
 
+def approx_cdf(expr, k=100):
+    """Produce a summary of the distribution of values.
+
+    .. include: _templates/experimental.rst
+
+    Notes
+    -----
+    This method returns a struct collecting two array: `values` and `ranks`. The
+    `values` array contains an ordered sample of values seen. The `ranks` array
+    is one longer, and contains the approximate ranks for the corresponding
+    values.
+
+    These represent a summary of the sorted list of values seen by the
+    aggregator. For example, values=[0,2,5,6,9] and ranks=[0,3,4,5,8,10]
+    represents the approximation [0,0,0,2,5,6,6,6,9,9], with the value
+    `values(i)` occupying indices `ranks(i)` (inclusive) to `ranks(i+1)`
+    (exclusive).
```

Comment:
For some context, I think of the `approx_cdf` aggregator as a low level thing that most users will never use directly. For a separate PR I've been adding approximate quantile aggregators and plotting methods, which are more user friendly endpoints. So I was writing this documentation with the superusers in mind who might want to build new use cases on top of this low level primitive.

Besides the fact that the ranks array is one longer than the values array, I think the intuition that the sketch is a list of approximate quantiles is subtly misleading. Not that it's wrong, exactly, but thinking that way tends to lead me to wrong conclusions.

Not sure if you will find this helpful, but it was clarifying for me to consider the extreme case where we collect only one sample. This implementation doesn't support that case, but if it did, the only thing it could do (and still be mathematically correct), is to collect a single uniform random sample, with weight N (the size of the data). Then the cdf would be `{values=[s], ranks=[0,N]}`.

The key property of the sketch is that, for any query `x`, we can return an unbiased estimate of the rank of `x`. In this case, we return 0 if `x≤s` and 1 if `x>s`. Note that this is in fact an unbiased estimate—if `s` is a uniform random sample, then `s<x` with probability equal to the quantile of `x`. If we flip it around and estimate the `q`th quantile, we will answer `s` for any `q`. In terms of the description I gave in the comment, that's because our sketch represents the stream that contains `N` copies of `s`. Your example gives the impression that in this case the sketch would try to find a sample close to the median, but that is not the case, nor are we providing any other quantile between 0 and 1 as our guess of the quantile of `s`.

I agree this could stand to be clarified, and I'll think more about the best thing to say. The best I can say right now is that this way of understanding the sketch is what I've had to keep falling back on when trying to figure out the right way to compute a quantile estimate, or how to draw a CDF plot.